Elasticsearch: Support for visualizing logs in Explore

marefr opened the pull request:
**What this PR does / why we need it**:
Support for visualizing logs in Explore using Elasticsearch.

Left to do for MVP:
- [x] Use field settings from datasource settings
- [x] Figure out the best way of sending query from query field to datasource.
- [x] Replace `context` property with something that says that the query is a logs query.
- [x] Logs query should be run when switching to logs mode even though it's empty.
- [x] Try use the `QueryField` without any highlighting or auto-complete to get styling and ctrl+Return/Return to run query. Need to look into why the input field is lagging when input text to it. If lag cannot be removed, lets go for a regular html input field instead with Return handler.
- [x] Add types for elasticsearch options to datasource in similar manner as Influx datasource

**Which issue(s) this PR fixes**:
Closes #16812 

**Special notes for your reviewer**:

marefr committed:
explore: try to use existing mode when switching datasource

If mode is Logs and datasource is changed to a datasource that supports
logs, then keep the Logs mode.

marefr committed:
elasticsearch: initial explore logs support

marefr committed:
fix tests

davkal commented:
Reduce scope for MVP:
- keep HTML input field for now, with Return handler
- clean up query handling, ambiguous query handling in the datasource, might need to probe on query type
- use field settings from datasource settings
- make sure empty (all) queries work, perhaps via `*`

6.3 beta is blocked by this.

kaydelaney commented:
Re: Logs query should be run when switching to logs mode even though it's empty.

The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
Perhaps a bit hacky though

kaydelaney committed:
Elasticsearch: Adds ElasticsearchOptions type
Updates tests accordingly

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -1,4 +1,15 @@
+  esVersion: number;
+  interval: string;
+  timeInterval: string;
+  maxConcurrentShardRequests: number;
```
This should probably be optional.

marefr commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -24,24 +28,32 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery> {
-    this.logMessageField = instanceSettings.jsonData.logMessageField || '';
-    this.logLevelField = instanceSettings.jsonData.logLevelField || '';
+    this.logMessageField = settingsData.logMessageField || '';
+    this.logLevelField = settingsData.logLevelField || '';
```
Awesome :+1:

marefr reviewed (commented):
Typings of options looks great! Left some comments. One thing to consider is to copy the typings for the `query` function from `DataSourceApi<>` which now will give you typed queries/targets and proper Promise as return value.

marefr commented:
> Re: Logs query should be run when switching to logs mode even though it's empty.
> 
> The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
> Perhaps a bit hacky though

@kaydelaney if we have a property on `ElasticsearchQuery` that says whether query is a logs query or not, that property can be added to hasNonEmptyQuery. What do you think?

kaydelaney commented:
> > Re: Logs query should be run when switching to logs mode even though it's empty.
> > The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
> > Perhaps a bit hacky though
> 
> @kaydelaney if we have a property on `ElasticsearchQuery` that says whether query is a logs query or not, that property can be added to hasNonEmptyQuery. What do you think?

Sounds good. 👍

kaydelaney committed:
Elasticsearch: Adds typing to query method

kaydelaney committed:
Elasticsearch: Makes maxConcurrentShardRequests optional

kaydelaney committed:
Explore: Allows empty query for elasticsearch datasource

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -0,0 +1,25 @@
+  settings?: any;
+}
+
+export interface ElasticsearchQuery extends DataQuery {
```
This should be used instead of the `ElasticsearchQuery` interface defined in `ElasticsearchQueryField.tsx`. Add missing fields from `ElasticsearchQueryField.tsx`.

kaydelaney replied:
So should the `ElasticsearchQuery` interface defined in `ElasticsearchQueryField.tsx` just be removed? (After adding the missing fields to the interface defined in `types.ts`)

marefr replied:
Yes, there shouldn't be anyone using that any longer.

marefr commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -253,38 +280,54 @@ export class ElasticDatasource {
-      const queryObj = this.queryBuilder.build(target, adhocFilters, queryString);
+
+      let queryObj;
+      if (target.context && target.context === 'explore') {
```
Think we can remove the `context` from the query now and only rely on `isLogsQuery` or does that screw up the check for empty query?

marefr reviewed (commented):
Great progress :+1: Left some comments that needs to be fixed.

kaydelaney committed:
Elasticsearch: Unifies ElasticsearchQuery interface definition
Removes check for context === 'explore'

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -0,0 +1,28 @@
+}
+
+export interface ElasticsearchQuery extends DataQuery {
+  context: string;
```
Is this still needed or can we remove it?

kaydelaney replied:
No longer needed I think. Good catch

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -0,0 +1,28 @@
+  alias?: string;
+  query?: string;
+  bucketAggs?: ElasticsearchAggregation[];
+  metricAggs?: ElasticsearchAggregation[];
```
`metricAggs` was a typo from my side I think and should be removed.

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -0,0 +1,28 @@
+  query?: string;
+  bucketAggs?: ElasticsearchAggregation[];
+  metricAggs?: ElasticsearchAggregation[];
+  metrics?: any[];
```
Possible to remove `any` and make it typed using `ElasticsearchAggregation[]`?

marefr commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -253,38 +279,54 @@ export class ElasticDatasource {
+
+      let queryObj;
+      if (target.isLogsQuery) {
+        console.log('Logs query!!!');
```
Remove comment :smile:

kaydelaney replied:
Oops! 😅

marefr reviewed (commented):
Changes looks great. Added some comments with request for minor code changes and cleanup.

Two things that doesn't work as expected is using the `Clear All` and clear query field row. They are actually clearing current query correctly, but the query isn't run at all which is the first problem. 
Second, if you after clear click `Run Query` nothing happens.

Let me know if you'll need more information provided to reproduce these problems.

kaydelaney committed:
Elasticsearch: Removes context property from ElasticsearchQuery interface
Adds field property
Removes console.log
Removes metricAggs property
Adds typing to metrics property

kaydelaney committed:
Elasticsearch: Runs default 'empty' query when 'clear all' button is pressed

marefr commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -0,0 +1,27 @@
+import { DataQuery, DataSourceJsonData } from '@grafana/ui';
+
+export interface ElasticsearchOptions extends DataSourceJsonData {
+  index: string;
```
This can be removed since it not in use anymore

marefr commented on public/app/plugins/datasource/elasticsearch/specs/datasource.test.ts:
```diff
@@ -77,9 +88,9 @@ describe('ElasticDatasource', function(this: any) {
-      });
+        database: '[asd-]YYYY.MM.DD',
+        jsonData: { interval: 'Daily', esVersion: 2 } as ElasticsearchOptions,
+      } as DataSourceInstanceSettings<ElasticsearchOptions>);
```
Would be nice with a test for "When issuing logs query"

marefr commented on public/app/plugins/datasource/elasticsearch/query_builder.ts:
```diff
@@ -367,4 +367,31 @@ export class ElasticQueryBuilder {
     return query;
   }
+
+  getLogsQuery(target, querystring) {
```
Would be nice with a test of this verifying that both documents and date histogram is requested.

marefr commented on public/app/plugins/datasource/elasticsearch/elastic_response.ts:
```diff
@@ -410,4 +412,140 @@ export class ElasticResponse {
     return { data: seriesList };
   }
+
+  getLogs(logMessageField?: string, logLevelField?: string): DataQueryResponse {
```
Would be nice with a test verifying that a response of both documents and date histogram aggregation can be properly parsed.

marefr commented on public/app/plugins/datasource/elasticsearch/components/ElasticsearchQueryField.tsx:
```diff
@@ -0,0 +1,94 @@
+          <div className="gf-form gf-form--grow flex-shrink-1">
+            <QueryField
+              additionalPlugins={this.plugins}
+              // cleanText={cleanText}
```
Remove commented code

marefr reviewed (commented):
All features now works as expected :+1: Some minor cleanup and adding some tests would be good before we merge this, see comments..

kaydelaney committed:
Elasticsearch: Removes index property from ElasticsearchOptions interface

kaydelaney committed:
Elasticsearch: Removes commented code from ElasticsearchQueryField.tsx

kaydelaney committed:
Elasticsearch: Adds comment warning usage of for...in to elastic_response.ts

kaydelaney committed:
Elasticsearch: adds tests related to log queries

marefr reviewed (commented):
Great work! Approved. Since I'm the author I cannot approve this PR. @kaydelaney you have to approve it
